### ISSUE TITLE
kore-repl: Solver timeout error

### ISSUE THREAD

**Issue by virgil-serbanuta** [kframework/kore#2418]
To reproduce:

Branch: https://github.com/virgil-serbanuta/verified-smart-contracts/tree/multisig.bug4/multisig/protocol-correctness/proof

```sh
cd multisig/protocol-correctness/proof
make execution.timestamp
kprove proof-perform-action-endpoint.k --debugger --debug-script kast.kscript
```
then, in the REPL: 
```
step 2
select 2
step 2
```
You can probably also use the kore files directly:
[kore-files.zip](https://github.com/kframework/kore/files/6029880/kore-files.zip)

The error is here:
[error.zip](https://github.com/kframework/kore/files/6029900/error.zip)

Sorry for not providing a smaller example, but it took me over a day to produce the current version, and almost any rule I delete, regardless of whether it's related to the actual proof, makes the error go away.

**Comment by MirceaS** [comment]
@virgil-serbanuta By rule do you mean one of the trusted claims in `proof-perform-action-endpoint.k`?

**Comment by virgil-serbanuta** [comment]
What I said also applies to the trusted claims, but I actually meant something that (for me, at least) is more interesting:

I don't remember exactly, but I think that deleting, say, almost any rule in the `PSEUDOCODE-TYPE-REFLECTION` module (https://github.com/virgil-serbanuta/verified-smart-contracts/blob/multisig.bug4/multisig/protocol-correctness/pseudocode.k#L182 ) will make the error go away, although they are not used in the semantics (I didn't actually try all of them). Also, I think that deleting any rule in the `EXECUTION-HELPERS` module should have the same effect (here I actually only tried deleting all the rules of a symbol at once, https://github.com/virgil-serbanuta/verified-smart-contracts/blob/multisig.bug4/multisig/protocol-correctness/proof/execution.k#L10 )

So, the following would probably be good examples:
```
  rule pListLen([.]) => 0
  rule pListLen([_:Expression, Es:ExpressionCSV]) => 1 +Int pListLen([Es])

  rule pListLen(_) >=Int 0 => true  [simplification, smt-lemma]

  rule cast(value(K:KItem) => K, _:ReflectionType)
```

**Comment by ttuegel** [comment]
We need to look at the equations the frontend generates for `isKResult`. The actual error here is related to a predicate that the SMT solver cannot handle. We noticed an unevaluated `isKResult` term there, which should be very simple to evaluate. The `isKResult` function is generated by the frontend according to some complex criteria, and it's possible that adding or removing rules that appear unrelated can change what equations are generated for that function.

**Comment by ttuegel** [comment]
There seems to be no difference in the `definition.kore` files with respect to `isKResult` when these rules are included or excluded, so that rules out my hypothesis.

**Comment by ttuegel** [comment]
@MirceaS 

1. If we remove one of these rules, do we get to the same configuration in the debugger (with or without the error)? Run `step 2`, `select 2`, `step 1` and verify that the configuration is the same.
2. Assuming the configurations are the same, run the debugger (with and without the rule) with `--solver-transcript` and compare the transcripts.

**Comment by MirceaS** [comment]
@virgil-serbanuta @ttuegel It looks like increasing the SMT timeout gets rid of the error.
I tried running `kprove --haskell-backend-command "kore-repl --smt-timeout 200" proof-perform-action-endpoint.k` and this seems to work consistently (in the sense that repl no longer fails).

Interestingly though, even if the rules are unrelated to what's being proven, the configuration just before that step differs on whether we kept some rule in or not. (I only experimented with removing ` rule cast(value(K:KItem) => K, _:ReflectionType)`)
